Speed up discovery tests

rosstimothy opened the pull request:
Removes the use of require.Never to reduce the test time by 10 seconds. The usage events should always be recorded synchronously during the reconciliation. By the time the waitForReconcile channel is written to the mock reporter will already have the stats from the reconciliation.

rosstimothy committed:
Speed up discovery tests

Removes the use of require.Never to reduce the test time by
10 seconds. The usage events should always be recorded synchronously
during the reconciliation. By the time the waitForReconcile channel
is written to the mock reporter will already have the stats from the
reconciliation.

rosstimothy commented:
I'd appreciate another look at this by those familiar with the discovery service. It looks like the require.Eventually/require.Never were added to address [flakes](https://github.com/gravitational/teleport/issues/31759) in the past. Though I can't quite see how the mockReporter isn't instrumented by Auth prior to returning from the SubmitUsageEvent RPC that happens during reconciliation.

marcoandredinis commented:
Just tested this and the flaky test is back after your changes.
Here's the command I used to test it:

```
CC=/opt/homebrew/bin/x86_64-unknown-linux-gnu-gcc GOOS=linux GOARCH=amd64 CGO_ENABLED=1 go test -c -race github.com/gravitational/teleport/lib/srv/discovery -ldflags '-w -s -I=/usr/lib64/ld-linux-x86-64.so.2' && docker run -it --cpus="0.4" --platform=linux/amd64 -v(pwd):/teleport -u1000 ubuntu:24.04 /teleport/discovery.test -test.count=20 -test.timeout=60m -test.v -test.failfast -test.run '^TestDiscoveryInCloudKube$'
```

The discovery service has a lot of async stuff, which makes it hard to get away without some sleeps here and there.
An alternative would be to convert this to `testing/synctest`.
It would take some time, but unless we add knobs in the DiscoveryService to trigger specific things for testing purposes, I'm not seeing another way of improving this.

Tener commented:
> The usage events should always be recorded synchronously during the reconciliation.

I think this isn't the case: `onKubeCreate` creates the cluster followed by emitting of events, but the tests are only waiting for the Kube cluster creation, which creates a race condition.

I think we could sync with `mockUsageReporter.AnonymizeAndSubmit` through a channel to eliminate polling.

rosstimothy commented:
@Tener @marcoandredinis I've added a callback for Kubernetes Cluster reconciliation that mimics the one used by databases and updated `TestDiscoveryInCloudKube` to use it. I've not been able to repro any failures for either test changed in this PR locally using a similar testing mechanism as the one Marco referenced above.
